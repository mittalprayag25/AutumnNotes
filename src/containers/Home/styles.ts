import {StyleSheet} from 'react-native';
import {sizes} from './../../constants/sizes';
import {colors} from './../../constants/colors';

const styles = StyleSheet.create({
  container: {
    height: '100%',
  },
  header: {
    marginBottom: sizes.space.md,
    fontSize: sizes.font.xl,
  },
  topContainer: {
    minHeight: 128,
  },
  editNote: {
    height: '20%',
    borderColor: colors.border,
    padding: sizes.space.md,
    borderWidth: 2,
    marginHorizontal: sizes.space.md,
  },
  navContainer: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  categoryLabel: {
    marginLeft: sizes.space.sm,
  },
  title: {
    fontSize: sizes.font.xl,
    alignSelf: 'flex-start',
    marginLeft: sizes.space.sm,
  },
  noRecords: {
    justifyContent: 'center',
    alignSelf: 'center',
  },
  categoryMenu: {
    marginLeft: 10,
  },
  menu: {
    alignItems: 'center',
    flexDirection: 'row',
    justifyContent: 'center',
  },
});

export default styles;
