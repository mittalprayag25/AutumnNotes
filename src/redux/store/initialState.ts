const initialState = {
  notes: {
    allNotes: [],
    singleNote: null,
  },
  categories: {
    categories: [],
  },
};

export default initialState;
