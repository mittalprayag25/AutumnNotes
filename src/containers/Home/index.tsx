import React, {useEffect, useState} from 'react';
import {Text, Image, View, TouchableOpacity} from 'react-native';
import {useDispatch, useSelector} from 'react-redux';
import actions from '../../redux/actions';
import CategoryList from './../../components/CategoryList';
import HeaderNav from './../../components/HeaderNav';
import {Routes} from '../../constants/NavigationUtils';
import {RootState} from '../../redux/reducers';
import images from './../../assets/index';
import styles from './styles';
import {ScrollView} from 'react-native-gesture-handler';

const Home = ({navigation}: any) => {
  const dispatch = useDispatch();
  const [categoryGroup, setAllCategoryGroup] = useState([]);

  const {notes}: any = useSelector((s: RootState) => s);

  useEffect(() => {
    let categoryArr = notes.allNotes.map(function (obj) {
      return obj.category;
    });

    categoryArr = categoryArr.filter(function (v, i) {
      return categoryArr.indexOf(v) == i;
    });
    const categoryGroupArr = [];
    for (let j = 0; j < categoryArr.length; j++) {
      const arr = notes.allNotes.filter(function (obj) {
        return obj.category === categoryArr[j];
      });
      const catObj: any = {};
      catObj.data = arr;
      catObj.category = categoryArr[j];
      categoryGroupArr.push(catObj);
    }

    setAllCategoryGroup(categoryGroupArr);
  }, [notes]);

  useEffect(() => {
    dispatch(actions.notes.allnotes());
    dispatch(actions.categories.allcategories());
  }, [dispatch]);

  const renderRightNav = () => (
    <>
      <TouchableOpacity
        onPress={() => navigation.navigate(Routes.CREATENODE)}
        testID="test-add">
        <Image source={images.notes.add} />
        <Text>Notes</Text>
      </TouchableOpacity>
      <Text
        testID="categories-text"
        style={styles.categoryMenu}
        onPress={() => navigation.navigate(Routes.CREATE_CATEGORY)}>
        Categories
      </Text>
    </>
  );

  return (
    <>
      <View style={styles.container}>
        <HeaderNav rightNav={renderRightNav()} title="Notes" />
        {categoryGroup.length > 0 ? (
          <ScrollView contentContainerStyle={{flexGrow: 1}}>
            {categoryGroup.map((data, index) => {
              return (
                data && (
                  <CategoryList
                    data={data}
                    navigation={navigation}
                    index={index}
                  />
                )
              );
            })}
          </ScrollView>
        ) : (
          <Text style={styles.noRecords} testID="no-record-found">
            No Records Found
          </Text>
        )}
      </View>
    </>
  );
};

export default Home;
