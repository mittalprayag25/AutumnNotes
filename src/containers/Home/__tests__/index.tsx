import {act, fireEvent, render} from '@testing-library/react-native';
import React from 'react';
import * as reactRedux from 'react-redux';
import {Provider} from 'react-redux';

import {
  createMockStore,
  mockNavigationWithParams,
} from '../../../__mocks__/index';
import Home from '../index';
import initialState from './../../../redux/store/initialState';

const mockedDispatch = jest.fn();

describe('Test Home', () => {
  const navigation = mockNavigationWithParams({});

  const useSelectorSpy = jest.spyOn(reactRedux, 'useSelector');
  const useDispatchSpy = jest.spyOn(reactRedux, 'useDispatch');

  beforeEach(() => {
    useDispatchSpy.mockReturnValue(mockedDispatch);
    useSelectorSpy.mockReturnValue({
      notes: {
        allNotes: [
          {
            value: 'some value',
            description: 'some description',
            category: 'some category',
          },
          {
            value: 'some value 2',
            description: 'some description 2',
            category: 'some category 2',
          },
        ],
      },
    });
  });

  const props = {
    navigation: navigation,
  };

  const changeState = {
    ...initialState,
  };
  it('should match snapshot', () => {
    const wrapper = render(
      <Provider store={createMockStore(changeState)}>
        <Home {...props} />
      </Provider>,
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
  it('should not show no record found', () => {
    const {queryByTestId} = render(
      <Provider store={createMockStore(changeState)}>
        <Home {...props} />
      </Provider>,
    );
    const textInput = queryByTestId('no-record-found');
    expect(textInput).toBeNull();
  });
});

describe('Test Home when all Notes is null', () => {
  const navigation = mockNavigationWithParams({});

  const useSelectorSpy = jest.spyOn(reactRedux, 'useSelector');
  const useDispatchSpy = jest.spyOn(reactRedux, 'useDispatch');

  beforeEach(() => {
    useDispatchSpy.mockReturnValue(mockedDispatch);
    useSelectorSpy.mockReturnValue({
      notes: {
        allNotes: [],
      },
    });
  });

  const props = {
    navigation: navigation,
  };

  const changeState = {
    ...initialState,
  };
  it('should match snapshot', () => {
    const wrapper = render(
      <Provider store={createMockStore(changeState)}>
        <Home {...props} />
      </Provider>,
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});

describe('Test Home with no notes', () => {
  const navigation = mockNavigationWithParams({});

  const useSelectorSpy = jest.spyOn(reactRedux, 'useSelector');
  const useDispatchSpy = jest.spyOn(reactRedux, 'useDispatch');

  beforeEach(() => {
    useDispatchSpy.mockReturnValue(mockedDispatch);
    useSelectorSpy.mockReturnValue({
      notes: {
        allNotes: [],
      },
    });
  });

  const props = {
    navigation: navigation,
  };

  const changeState = {
    ...initialState,
  };
  it('should match snapshot', () => {
    const wrapper = render(
      <Provider store={createMockStore(changeState)}>
        <Home {...props} />
      </Provider>,
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should show no record found', () => {
    const {getByTestId} = render(
      <Provider store={createMockStore(changeState)}>
        <Home {...props} />
      </Provider>,
    );
    const textInput = getByTestId('no-record-found');
    expect(textInput).not.toBeNull();
  });
  it('should Navigate to new Note Page', () => {
    const {getByTestId} = render(
      <Provider store={createMockStore(changeState)}>
        <Home {...props} />
      </Provider>,
    );
    const add = getByTestId('test-add');
    act(() => {
      fireEvent(add, 'press');
    });
    expect(navigation.navigate).toHaveBeenCalledTimes(1);
  });
  it('Should open category page', () => {
    const {getByTestId} = render(
      <Provider store={createMockStore(changeState)}>
        <Home {...props} />
      </Provider>,
    );
    const category = getByTestId('categories-text');
    act(() => {
      fireEvent(category, 'press');
    });
    expect(navigation.navigate).toHaveBeenCalledTimes(2);
  });
});
