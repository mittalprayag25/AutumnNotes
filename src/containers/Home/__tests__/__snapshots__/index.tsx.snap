// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Home should match snapshot 1`] = `
<View
  style={
    Object {
      "height": "100%",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "borderBottomWidth": 1,
        "borderColor": "#abc",
        "flexDirection": "row",
        "height": 50,
        "justifyContent": "space-between",
        "paddingHorizontal": 8,
        "width": "100%",
      }
    }
  >
    <Text
      style={
        Object {
          "alignItems": "center",
          "fontSize": 20,
          "fontWeight": "bold",
          "left": "40%",
          "position": "relative",
        }
      }
    >
      Notes
    </Text>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="test-add"
      >
        <Image
          source={
            Object {
              "testUri": "../../../src/assets/add.png",
            }
          }
        />
        <Text>
          Notes
        </Text>
      </View>
      <Text
        onPress={[Function]}
        style={
          Object {
            "marginLeft": 10,
          }
        }
        testID="categories-text"
      >
        Categories
      </Text>
    </View>
  </View>
  <RCTScrollView
    collapsable={false}
    contentContainerStyle={
      Object {
        "flexGrow": 1,
      }
    }
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
  >
    <View>
      <View
        style={
          Object {
            "backgroundColor": "#eeeaaa",
            "marginVertical": 8,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 20,
              "fontWeight": "bold",
              "marginLeft": 8,
            }
          }
        >
          some category
        </Text>
        <RCTScrollView
          data={
            Array [
              Object {
                "category": "some category",
                "description": "some description",
                "value": "some value",
              },
            ]
          }
          getItem={[Function]}
          getItemCount={[Function]}
          keyExtractor={[Function]}
          onContentSizeChange={[Function]}
          onLayout={[Function]}
          onMomentumScrollBegin={[Function]}
          onMomentumScrollEnd={[Function]}
          onScroll={[Function]}
          onScrollBeginDrag={[Function]}
          onScrollEndDrag={[Function]}
          removeClippedSubviews={false}
          renderItem={[Function]}
          scrollEventThrottle={50}
          stickyHeaderIndices={Array []}
          style={Object {}}
          viewabilityConfigCallbackPairs={Array []}
        >
          <View>
            <View
              onLayout={[Function]}
              style={null}
            >
              <View
                style={
                  Object {
                    "flexDirection": "row",
                  }
                }
              >
                <View
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "alignItems": "flex-start",
                      "backgroundColor": "#eeeee4",
                      "borderRadius": 8,
                      "height": 128,
                      "margin": 8,
                      "opacity": 1,
                      "padding": 8,
                      "width": "45%",
                    }
                  }
                  testID="note-detail-item-undefined"
                >
                  <Text
                    style={
                      Object {
                        "fontSize": 16,
                      }
                    }
                  >
                    undefined
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </RCTScrollView>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "#eeeaaa",
            "marginVertical": 8,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 20,
              "fontWeight": "bold",
              "marginLeft": 8,
            }
          }
        >
          some category 2
        </Text>
        <RCTScrollView
          data={
            Array [
              Object {
                "category": "some category 2",
                "description": "some description 2",
                "value": "some value 2",
              },
            ]
          }
          getItem={[Function]}
          getItemCount={[Function]}
          keyExtractor={[Function]}
          onContentSizeChange={[Function]}
          onLayout={[Function]}
          onMomentumScrollBegin={[Function]}
          onMomentumScrollEnd={[Function]}
          onScroll={[Function]}
          onScrollBeginDrag={[Function]}
          onScrollEndDrag={[Function]}
          removeClippedSubviews={false}
          renderItem={[Function]}
          scrollEventThrottle={50}
          stickyHeaderIndices={Array []}
          style={Object {}}
          viewabilityConfigCallbackPairs={Array []}
        >
          <View>
            <View
              onLayout={[Function]}
              style={null}
            >
              <View
                style={
                  Object {
                    "flexDirection": "row",
                  }
                }
              >
                <View
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "alignItems": "flex-start",
                      "backgroundColor": "#eeeee4",
                      "borderRadius": 8,
                      "height": 128,
                      "margin": 8,
                      "opacity": 1,
                      "padding": 8,
                      "width": "45%",
                    }
                  }
                  testID="note-detail-item-undefined"
                >
                  <Text
                    style={
                      Object {
                        "fontSize": 16,
                      }
                    }
                  >
                    undefined
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </RCTScrollView>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Test Home when all Notes is null should match snapshot 1`] = `
<View
  style={
    Object {
      "height": "100%",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "borderBottomWidth": 1,
        "borderColor": "#abc",
        "flexDirection": "row",
        "height": 50,
        "justifyContent": "space-between",
        "paddingHorizontal": 8,
        "width": "100%",
      }
    }
  >
    <Text
      style={
        Object {
          "alignItems": "center",
          "fontSize": 20,
          "fontWeight": "bold",
          "left": "40%",
          "position": "relative",
        }
      }
    >
      Notes
    </Text>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="test-add"
      >
        <Image
          source={
            Object {
              "testUri": "../../../src/assets/add.png",
            }
          }
        />
        <Text>
          Notes
        </Text>
      </View>
      <Text
        onPress={[Function]}
        style={
          Object {
            "marginLeft": 10,
          }
        }
        testID="categories-text"
      >
        Categories
      </Text>
    </View>
  </View>
  <Text
    style={
      Object {
        "alignSelf": "center",
        "justifyContent": "center",
      }
    }
    testID="no-record-found"
  >
    No Records Found
  </Text>
</View>
`;

exports[`Test Home with no notes should match snapshot 1`] = `
<View
  style={
    Object {
      "height": "100%",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "borderBottomWidth": 1,
        "borderColor": "#abc",
        "flexDirection": "row",
        "height": 50,
        "justifyContent": "space-between",
        "paddingHorizontal": 8,
        "width": "100%",
      }
    }
  >
    <Text
      style={
        Object {
          "alignItems": "center",
          "fontSize": 20,
          "fontWeight": "bold",
          "left": "40%",
          "position": "relative",
        }
      }
    >
      Notes
    </Text>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="test-add"
      >
        <Image
          source={
            Object {
              "testUri": "../../../src/assets/add.png",
            }
          }
        />
        <Text>
          Notes
        </Text>
      </View>
      <Text
        onPress={[Function]}
        style={
          Object {
            "marginLeft": 10,
          }
        }
        testID="categories-text"
      >
        Categories
      </Text>
    </View>
  </View>
  <Text
    style={
      Object {
        "alignSelf": "center",
        "justifyContent": "center",
      }
    }
    testID="no-record-found"
  >
    No Records Found
  </Text>
</View>
`;
